### PR TITLE
Use vanilla llvm

### PR DESCRIPTION
**Pull request by vgvassilev**
Enable ROOT's builtin_llvm=Off

This patch allows ROOT to be built against compatible external llvm (5.0 or 5.0.1). Note that we still need to build clang (eg. we require builtin_clang=On) due to the ROOT-specific patches which are not yet upstream.

Since we have externally installed llvm, we configure and build clang as a standalone project. The configuration relies on finding llvm-config-5.0 and uses an adapted version of the standard clang standalone build procedure.

Clang provides dependencies such as FileCheck and not which are used by cling's testsuite and are not being installed with the standard llvm package.

Cling (which depends on llvm and clang) is built as a clang tool to avoid unresolved dependencies to clang and complicating further the already complicated cmake setup.
    
This patch intends a minimal change and follows the initial (suboptimal) design to configure and build llvm, clang and cling as part of ROOT. An ultimate solution would be to have llvm, clang and cling built as separate standalone projects (following the recommended way by the LLVM cmake developers).

This should reduce the compile times by approx 20% (if we have llvm-5.0-dev installed) and make cross compilation of ROOT a little easier.

### CODE REVIEW

**Comment by vgvassilev** [comment]
@phsft-bot build!

File: core/dictgen/CMakeLists.txt [hunk 1]
```diff
@@ -13,9 +13,13 @@ ROOT_GLOB_SOURCES(sources ${CMAKE_CURRENT_SOURCE_DIR}/src/*.cxx)
...
         ${CMAKE_CURRENT_SOURCE_DIR}/../foundation/res
-        ${CMAKE_CURRENT_SOURCE_DIR}/../clingutils/res
-        ${LLVM_INCLUDE_DIRS}
-        ${CLING_INCLUDE_DIRS})
+        ${CMAKE_CURRENT_SOURCE_DIR}/../clingutils/res)
+include_directories(${CLING_INCLUDE_DIRS})
+# Register the llvm include directories after clangs. This instructs the compiler to resolve
+# headers from our builtin clang. That's an issue when we are building with bultin_llvm=Off
+# and we have installed clang headers, too.
+include_directories(SYSTEM ${CLANG_INCLUDE_DIRS} ${LLVM_INCLUDE_DIRS})
```
Comment (amadio) [inline core/dictgen/CMakeLists.txt]: You may want to use `BEFORE` keyword when adding `${CLANG_INCLUDE_DIRS}`.

File: core/metacling/src/CMakeLists.txt [hunk 2]
```diff
@@ -3,10 +3,14 @@
...
 
-include_directories(${LLVM_INCLUDE_DIRS} ${CLING_INCLUDE_DIRS})
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../res
         ${CMAKE_CURRENT_SOURCE_DIR}/../../clingutils/res
         ${CMAKE_CURRENT_SOURCE_DIR}/../../foundation/res)
+include_directories(${CLING_INCLUDE_DIRS})
+# Register the llvm include directories after clangs. This instructs the compiler to resolve
+# headers from our builtin clang. That's an issue when we are building with bultin_llvm=Off
+# and we have installed clang headers, too.
+include_directories(SYSTEM ${CLANG_INCLUDE_DIRS} ${LLVM_INCLUDE_DIRS})
```
Comment (amadio) [inline core/metacling/src/CMakeLists.txt]: This is a repetition of what's in `core/dictgen`, you may not need this at all if you use `target_include_directories()` for the dependencies.

File: interpreter/CMakeLists.txt [hunk 3]
```diff
@@ -80,7 +99,14 @@ set(LLVM_TARGETS_TO_BUILD ${ROOT_CLING_TARGET} CACHE STRING "Semicolon-separated
...
   set(CLING_INCLUDE_TESTS ON CACHE BOOL "" FORCE)
-  ROOT_ADD_TEST(clingtest-check-cling COMMAND ${CMAKE_COMMAND} --build ${CMAKE_BINARY_DIR} --target check-cling)
+  # The path to cling passed through environment variable only relevant when building
+  # against external llvm. In that case, cling's testsuite cannot deduce the binary
+  # of cling relatively to the llvm tools folder.
+  if (NOT builtin_llvm)
+    set(CLINGTEST_EXECUTABLE CLING=${CMAKE_CURRENT_BINARY_DIR}/llvm/src/${CMAKE_CFG_INTDIR}/bin/cling)
+  endif()
+  ROOT_ADD_TEST(clingtest-check-cling COMMAND ${CMAKE_COMMAND} --build ${CMAKE_BINARY_DIR} --target check-cling
+                                       ENVIRONMENT ${CLINGTEST_EXECUTABLE})
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: Just curiosity, but why is this needed? Could you use `$<TARGET_FILE:cling>` or similar in `check-cling` target to avoid this?

File: interpreter/CMakeLists.txt [hunk 4]
```diff
@@ -134,24 +160,243 @@ set_directory_properties(PROPERTIES INCLUDE_DIRECTORIES "")
...
+    add_subdirectory(llvm/src EXCLUDE_FROM_ALL)
+  endif()
+
+  #---Make LLVM compilation flags public---------------------------------------------------------------
+  # Extract the compilation flags from LLVM and make them public to the
+  # rest of ROOT so that we can compile against LLVM with matching flags.
+
+  # LLVM doesn't really give us a API to get this with an in-source build
+  # so we just use the normal way of doing this and read the llvm directory
+  # compilation properties.
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: This is strange, the flags should come with the targets already, no? If you use `target_link_libraries()`, any flags needed will be automatically appended (i.e. include directories and compile flags). This seems a bit hacky.

File: interpreter/CMakeLists.txt [hunk 5]
```diff
@@ -134,24 +160,243 @@ set_directory_properties(PROPERTIES INCLUDE_DIRECTORIES "")
...
+  set(ROOT_LLVM_FLAGS "${LLVM_DEFS}" CACHE STRING "" FORCE)
 else()
-  add_subdirectory(llvm/src EXCLUDE_FROM_ALL)
-endif()
+  if (NOT builtin_clang)
+    message(FATAL_ERROR "ROOT cannot work with builtin clang. Please use -Dbuiltin_clang=Off to turn it off.")
+  endif()
+  # Rely on llvm-config.
+  set(CONFIG_OUTPUT)
+  find_program(LLVM_CONFIG NAMES "llvm-config-5.0" "llvm-config")
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: Why not use LLVM's CMake modules? Macs don't have `llvm-config` from XCode, for example, although I understand that this is probably Linux-only since the version of LLVM has to almost match what's in ROOT.
On my machine, the file `LLVMConfig.cmake` has the flags returned by `llvm-config --cxxflags`, for example, as well as paths to various directories.

File: interpreter/CMakeLists.txt [hunk 6]
```diff
@@ -134,24 +160,243 @@ set_directory_properties(PROPERTIES INCLUDE_DIRECTORIES "")
...
+      CACHE BOOL "Enable assertions")
+    # Assertions should follow llvm-config's.
+    mark_as_advanced(LLVM_ENABLE_ASSERTIONS)
+  endif()
+
+  set(LLVM_TOOLS_BINARY_DIR ${TOOLS_BINARY_DIR} CACHE PATH "Path to llvm/bin")
+  set(LLVM_LIBRARY_DIR ${LIBRARY_DIR} CACHE PATH "Path to llvm/lib")
+  set(LLVM_MAIN_INCLUDE_DIR ${INCLUDE_DIR} CACHE PATH "Path to llvm/include")
+  set(LLVM_BINARY_DIR ${LLVM_OBJ_ROOT} CACHE PATH "Path to LLVM build tree")
+  set(LLVM_MAIN_SRC_DIR ${MAIN_SRC_DIR} CACHE PATH "Path to LLVM source tree")
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: I think you would get all these paths with a simple `find_package(LLVM)`, no? These are all in the `LLVMConfig.cmake` installed by LLVM on my machine.

File: interpreter/CMakeLists.txt [hunk 7]
```diff
@@ -134,24 +160,243 @@ set_directory_properties(PROPERTIES INCLUDE_DIRECTORIES "")
...
-string (REPLACE ";" " -D" LLVM_DEFS ";${LLVM_DEFS}")
-# CACHE variable that contains the additional flags that LLVM needs.
-set(ROOT_LLVM_FLAGS "${LLVM_DEFS}" CACHE STRING "" FORCE)
+  include(AddLLVM)
+  include(TableGen)
+  include(HandleLLVMOptions)
+  include(VersionFromVCS)
+
+  set(PACKAGE_VERSION "${LLVM_PACKAGE_VERSION}")
+  if (NOT PACKAGE_VERSION VERSION_EQUAL 5.0)
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: Your PR says it also works with 5.0.1, is this true here if `PACKAGE_VERSION` is 5.0.1?

File: interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp [hunk 8]
```diff
@@ -65,11 +65,13 @@ static void LLVMErrorHandler(void *UserData, const std::string &Message,
...
 }
 
-#ifdef LINK_POLLY_INTO_TOOLS
+#if 0
+#if defined(LINK_POLLY_INTO_TOOLS) && LINK_POLLY_INTO_TOOLS != 0
 namespace polly {
 void initializePollyPasses(llvm::PassRegistry &Registry);
 }
 #endif
+#endif
```
Comment (amadio) [inline interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp]: Doesn't this affect the build with `builtin_llvm=ON`?

**Comment by amadio** [comment]
Hi @vgvassilev, should I try to test this manually both with `builtin_llvm=ON` and `OFF`? Or is there already a build for testing it that you are using?

**Comment by amadio** [comment]
BTW, Gentoo users will really appreciate this change, since once this becomes possible I will immediately move to using external LLVM there. :smiley:

File: interpreter/CMakeLists.txt (hunk 3 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: We did not need this before, because we knew where to find cling, relative to the `llvm_tools_dir`. Having builtin_llvm=Off sets that variable to something else and we cannot find back the cling binary.

Of course we can probably find a hackish solution but I did not find anything that I liked. Currently the `check-cling` target is only available when configuring with -Dclingtest=On and if we have -Dbuiltin_llvm=Off we will need to provide the path to the cling binary if we want to dial the target outside of roottest.

File: interpreter/CMakeLists.txt (hunk 5 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: This did not work well for llvm. One of the reasons is that clang would also include AddLLVM, which will include `LLVMConfig.cmake`.

My initial implementation was this but for some reason I had missing dependencies and dependency cycles. Then I decided to copy what clang's main CMakeLists.txt file is doing...

File: interpreter/CMakeLists.txt (hunk 6 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: See above. Did not work.

File: interpreter/CMakeLists.txt (hunk 7 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: Yes, that is correct. LLVM 5.0.1 is a patch release and is supposed to be ABI compatible (to my surprise ;). On systems such as Ubuntu if you install llvm-5.0-dev you get 5.0.1 actually, so it has to work with 5.0.1.

File: interpreter/CMakeLists.txt (hunk 5 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: What if we just add the clang subdirectory and let it handle how to find and configure LLVM for its needs? Would that work, maybe?

File: interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp (hunk 8 above)
Comment (vgvassilev) [reply interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp]: No, because the our llvm config undefines polly. However the standard llvm install ships polly libraries and we end up having something like `#define LINK_POLLY_INTO_TOOLS` instead of `#define LINK_POLLY_INTO_TOOLS 0`. That a bug in clang.

File: interpreter/CMakeLists.txt (hunk 7 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: No, what I mean is if `PACKAGE_VERSION VERSION_EQUAL 5.0` evaluates to true even is the version is `5.0.1` and not `5.0`. I guess it does then?

File: core/clingutils/CMakeLists.txt [hunk 9]
```diff
@@ -10,8 +10,13 @@ ROOT_GLOB_HEADERS(headers ${CMAKE_CURRENT_SOURCE_DIR}/inc/*.h)
...
 set(ClingUtils_dict_headers ${headers} PARENT_SCOPE)
 
-include_directories(${LLVM_INCLUDE_DIRS} ${CLING_INCLUDE_DIRS})
+
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/res ${CMAKE_CURRENT_SOURCE_DIR}/../foundation/res)
+include_directories(${CLING_INCLUDE_DIRS})
+# Register the llvm include directories after clangs. This instructs the compiler to resolve
+# headers from our builtin clang. That's an issue when we are building with bultin_llvm=Off
+# and we have installed clang headers, too.
+include_directories(SYSTEM ${CLANG_INCLUDE_DIRS} ${LLVM_INCLUDE_DIRS})
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply core/clingutils/CMakeLists.txt]: Sure. I'd prefer to do this outside of this PR.

File: interpreter/CMakeLists.txt (hunk 7 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: Yes, that's the intent.

File: interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp (hunk 8 above)
Comment (amadio) [reply interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp]: Ok, thanks for the explanation. Is this already fixed upstream? We could just add `#undef LINK_POLLY_INTO_TOOLS` at the beginning of the file, then, right?

**Comment by vgvassilev** [comment]
I was running some compilation benchmarks on a haswell machine and it looks like in `-j12` mode we are not scaling well even if llvm is out of our way. I suspect we still have a bridge in the graph around cling/rootcling.

I.e. I was expecting better compilation speedup...

File: cmake/modules/CheckCompiler.cmake [hunk 10]
```diff
@@ -80,21 +80,6 @@ else()
...
-endif()
-if(NOT CMAKE_BUILD_TYPE)
-  set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Choose the type of build, options are: Release, MinSizeRel, Debug, RelWithDebInfo, Optimized." FORCE)
-endif()
-string(TOUPPER "${CMAKE_BUILD_TYPE}" uppercase_CMAKE_BUILD_TYPE)
-string(TOUPPER "${CMAKE_CONFIGURATION_TYPES}" uppercase_CMAKE_CONFIGURATION_TYPES)
-if(NOT "${uppercase_CMAKE_CONFIGURATION_TYPES}" MATCHES "${uppercase_CMAKE_BUILD_TYPE}")
-  message(FATAL_ERROR "CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE} needs to be one of known build types: ${CMAKE_CONFIGURATION_TYPES}")
-endif()
-
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply cmake/modules/CheckCompiler.cmake]: That's PR #1633.

File: core/dictgen/CMakeLists.txt (hunk 1 above)
Comment (vgvassilev) [reply core/dictgen/CMakeLists.txt]: I am not sure I get your idea. I need `CLANG_INCLUDE_DIRS` to show up before `LLVM_INCLUDE_DIRS`. IIUC, if I want to use `BEFORE` I will have to have at least two lines. Can you paste a code snippet here and I'll be happy to use it.

File: core/metacling/src/CMakeLists.txt (hunk 2 above)
Comment (vgvassilev) [reply core/metacling/src/CMakeLists.txt]: Sigh... yes. I'd prefer to do this outside this. I want this PR to introduce minimal incremental changes enabling external llvm. Otherwise, I will have too many moving targets :(

File: interpreter/CMakeLists.txt (hunk 4 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: That's legacy code, I moved it and indented it, that's why it shows up here. I do not know how to solve this at the moment.

File: interpreter/CMakeLists.txt (hunk 5 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: It'd misplace the binaries. The copy-pasted code is not 1:1, it has a few things modified. Namely, it'd install 'private' cling/clang libs and binaries in `$ROOTSYS/bin/`. That would make things inconsistent if people build with `-Dbuiltin_llvm=On|Off`.

File: interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp (hunk 8 above)
Comment (vgvassilev) [reply interpreter/llvm/src/tools/clang/tools/driver/cc1_main.cpp]: This is not yet fixed upstream. I will have to come up with a patch. `#undef`-ing would work (that's what our llvm-config does when polly is disabled) but it would be inconsistent with the way clang handles such options. My patch (except for the `if 0` does what other similar cases are doing).

File: core/clingutils/CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (amadio) [reply core/clingutils/CMakeLists.txt]: It would be nice to do it using `target_include_directories()` directly, but if you want to do it later, it's fine. However, since this is repeated in multiple subdirectories, it might make more sense to just do it one directory up instead.

File: interpreter/CMakeLists.txt (hunk 4 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: Ok, let's leave it for later than.

File: interpreter/CMakeLists.txt (hunk 5 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: I see. Ok, then.

File: core/clingutils/CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (vgvassilev) [reply core/clingutils/CMakeLists.txt]: Sure, that's a good idea.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
ping... if there are no other comments I'd like to merge this in.

File: cmake/modules/RootBuildOptions.cmake [hunk 11]
```diff
@@ -272,7 +274,7 @@ endif()
 
 #---Apply minimal or gminimal------------------------------------------------------------------
 foreach(opt ${root_build_options})
-  if(NOT opt MATCHES "thread|cxx11|cling|builtin_llvm|builtin_ftgl|explicitlink")
+  if(NOT opt MATCHES "thread|cxx11|cling|builtin_llvm|builtin_clang|builtin_ftgl|explicitlink")
```
Comment (amadio) [inline cmake/modules/RootBuildOptions.cmake]: Why set `builtin_clang` to `OFF` here if ROOT doesn't work when it's `OFF`?

File: interpreter/CMakeLists.txt [hunk 12]
```diff
@@ -134,24 +160,243 @@ set_directory_properties(PROPERTIES INCLUDE_DIRECTORIES "")
...
+    endif()
+  endforeach()
+  set(LLVM_VERSION "${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR}.${LLVM_VERSION_PATCH}" PARENT_SCOPE)
+else()
+  set(LLVM_INCLUDE_DIRS ${LLVM_MAIN_INCLUDE_DIR}
+    CACHE STRING "System LLVM include directories."
+    )
+  # We cannot include LLVMConfig.cmake here. We reply on builtin clang to discover the available
+  # llvm version and set it in clang version.
+  # FIXME: This is a very ugly workaround. Is there a better way to do this?
```
Comment (amadio) [inline interpreter/CMakeLists.txt]: Maybe you should consider addressing this and other `FIXME` comments before merging this.

**Comment by amadio** [comment]
Hi Vassil. It seems this PR still has some hacks and commented out code that could be cleaned up before merging.

File: interpreter/CMakeLists.txt (hunk 12 above)
Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: That's a reminder that we are being suboptimal here. I am unaware of how to fix it at that point in time, even if I had the spare cycles.

**Comment by vgvassilev** [comment]
@amadio, the commented code is intentional: it is part of a copy-paste from clang (as we discussed before) and those parts were commented out to adapt it to fit ROOT requirements.

File: interpreter/CMakeLists.txt (hunk 12 above)
Comment (amadio) [reply interpreter/CMakeLists.txt]: Fine, but then please let me take a look at this (since I am already working on improving our current system for externals/builtins) to try to find a better solution instead of merging something suboptimal.

Comment (vgvassilev) [reply interpreter/CMakeLists.txt]: I'd prefer to merge this because it is blocking a few things. Then once you find a better solution feel free to change this code there.

Comment (amadio) [reply interpreter/CMakeLists.txt]: For this one time, sure. Go ahead. But I will not accept merging of hacky changes to the build system in the future anymore.

**Comment by amadio** [comment]
I am now testing this, and see the following error:
```
Scanning dependencies of target G__Core
[ 67%] Generating G__Core.cxx, ../../lib/libCore.rootmap
In file included from input_line_1:1:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/new:40:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/exception:142:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/exception_ptr.h:38:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/cxxabi_init_exception.h:38:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
In file included from input_line_4:2:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/stdlib.h:30:
In file included from /home/amadio/build/external-llvm/etc/cling/lib/clang/5.0.1/include/stdlib.h:8:
/usr/include/stdlib.h:31:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
Error: Error loading the default header files.
gmake[2]: *** [core/base/CMakeFiles/G__Core.dir/build.make:435: core/base/G__Core.cxx] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:8972: core/base/CMakeFiles/G__Core.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 67%] Built target Cling
gmake: *** [Makefile:152: all] Error 2
```

**Comment by amadio** [comment]
Here is the verbose call to rootcling:
```
cd /home/amadio/build/external-llvm/core/base && /usr/bin/cmake -E env LD_LIBRARY_PATH=/home/amadio/build/external-llvm/lib:/usr/bin/../lib: /home/amadio/build/external-llvm/core/rootcling_stage1/src/rootcling_stage1 -v2 -f G__Core.cxx -s /home/amadio/build/external-llvm/lib/libCore.so -excludePath /home/amadio/src/root -excludePath /home/amadio/build/external-llvm -rml libCore.so -rmf /home/amadio/build/external-llvm/lib/libCore.rootmap -DSYSTEM_TYPE_unix -I/home/amadio/src/root -I/home/amadio/src/root/interpreter/cling/include -I/home/amadio/build/external-llvm/include -I/home/amadio/src/root/core/base/res -I/home/amadio/src/root/core/base/../foundation/res -I/usr/include TApplication.h TApplicationImp.h TAtt3D.h TAttAxis.h TAttBBox.h TAttBBox2D.h TAttFill.h TAttLine.h TAttMarker.h TAttPad.h TAttText.h TBase64.h TBenchmark.h TBrowser.h TBrowserImp.h TBuffer.h TBuffer3D.h TBuffer3DTypes.h TCanvasImp.h TColor.h TColorGradient.h TContextMenu.h TContextMenuImp.h TControlBarImp.h TDatime.h TDirectory.h TEnv.h TError.h TException.h TExec.h TFileCollection.h TFileInfo.h TFolder.h TGuiFactory.h TInetAddress.h TInspectorImp.h TMD5.h TMacro.h TMathBase.h TMemberInspector.h TMessageHandler.h TNamed.h TObjString.h TObject.h TObjectSpy.h TPRegexp.h TParameter.h TPluginManager.h TPoint.h TProcessID.h TProcessUUID.h TQClass.h TQCommand.h TQConnection.h TQObject.h TROOT.h TRedirectOutputGuard.h TRef.h TRefCnt.h TRegexp.h TRemoteObject.h TRootIOCtor.h TStopwatch.h TStorage.h TString.h TStringLong.h TStyle.h TSysEvtHandler.h TSystem.h TSystemDirectory.h TSystemFile.h TTask.h TThreadSlots.h TTime.h TTimeStamp.h TTimer.h TUUID.h TUri.h TUrl.h TVersionCheck.h TVirtualAuth.h TVirtualFFT.h TVirtualGL.h TVirtualMonitoring.h TVirtualMutex.h TVirtualPS.h TVirtualPad.h TVirtualPadEditor.h TVirtualPadPainter.h TVirtualPerfStats.h TVirtualQConnection.h TVirtualRWMutex.h TVirtualTableInterface.h TVirtualViewer3D.h TVirtualX.h Rtypes.h GuiTypes.h MessageTypes.h KeySymbols.h Buttons.h strlcpy.h snprintf.h TArray.h TArrayC.h TArrayD.h TArrayF.h TArrayI.h TArrayL.h TArrayL64.h TArrayS.h TBits.h TBtree.h TClassTable.h TClonesArray.h TCollection.h TCollectionProxyInfo.h TExMap.h THashList.h THashTable.h TIterator.h TList.h TMap.h TObjArray.h TObjectTable.h TOrdCollection.h TRefArray.h TRefTable.h TSeqCollection.h TSortedList.h TVirtualCollectionProxy.h ROOT/TSeq.hxx ESTLType.h RStringView.h TClassEdit.h ROOT/RArrayView.hxx ROOT/RMakeUnique.hxx ROOT/RNotFn.hxx ROOT/RStringView.hxx ROOT/TypeTraits.hxx ROOT/rhysd_array_view.hxx TUnixSystem.h TClingRuntime.h root_std_complex.h TBaseClass.h TClass.h TClassGenerator.h TClassMenuItem.h TClassRef.h TClassStreamer.h TDataMember.h TDataType.h TDictAttributeMap.h TDictionary.h TEnum.h TEnumConstant.h TFileMergeInfo.h TFunction.h TFunctionTemplate.h TGenericClassInfo.h TGlobal.h TInterpreter.h TInterpreterValue.h TIsAProxy.h TListOfDataMembers.h TListOfEnums.h TListOfEnumsWithLock.h TListOfFunctionTemplates.h TListOfFunctions.h TMemberStreamer.h TMethod.h TMethodArg.h TMethodCall.h TProtoClass.h TRealData.h TSchemaHelper.h TSchemaRule.h TSchemaRuleSet.h TStatusBitsChecker.h TStreamer.h TStreamerElement.h TToggle.h TToggleGroup.h TVirtualIsAProxy.h TVirtualRefProxy.h TVirtualStreamerInfo.h Getline.h /home/amadio/src/root/core/base/inc/LinkDef.h
In file included from input_line_1:1:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/new:40:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/exception:142:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/exception_ptr.h:38:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/cxxabi_init_exception.h:38:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
In file included from input_line_4:2:
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/stdlib.h:30:
In file included from /home/amadio/build/external-llvm/etc/cling/lib/clang/5.0.1/include/stdlib.h:8:
/usr/include/stdlib.h:31:10: fatal error: 'stddef.h' file not found
#include <stddef.h>
         ^~~~~~~~~~
Error: Error loading the default header files.
```

**Comment by vgvassilev** [comment]
Resolved. @amadio will push a fix.
